algorithms/kernel/kmeans/oneapi/cl_kernels/kmeans_cl_kernels_2.cl

```diff
@@ -0,0 +1,448 @@
+                                          __global algorithmFPType * distances, int K, int P) {
+        const int global_id_0  = get_global_id(0);
+        const int local_id_1   = get_local_id(1);
+        const int local_size_1 = get_local_size(1);
+        const int j            = global_id_0 * P + local_id_1;
+        algorithmFPType res;
+        __global const algorithmFPType * p = (__global const algorithmFPType *)&data[j];
+        res                                = 0.5 * data[j] * data[j];
+        algorithmFPType val                = sub_group_reduce_add(res);
+        dataSq[global_id_0]                = val;
+        for (int i = local_id_1; i < K; i += local_size_1) distances[global_id_0 + i] = val;
+    }
+
+    __kernel void init_distances(__global const algorithmFPType * centroidsSq, __global algorithmFPType * distances, int N, int K) {
+        const int global_id_0 = get_global_id(0);
+        const int global_id_1 = get_global_id(1);
+
+        if (global_id_0 < N)
+        {
+            distances[global_id_0 + global_id_1 * N] = centroidsSq[global_id_1];
+        }
+    }
+
+    __kernel void reduce_assignments(__global const algorithmFPType * distances, __global int * assignments, __global algorithmFPType * mindistances,
+                                     int N, int K) {
+        const int global_id = get_global_id(0);
+        const int size      = get_local_size(1);
+        const int local_id  = get_local_id(1);
+        int numgrp          = K / size;
+        const int rem       = K % size;
+        if (rem > 0) numgrp++;
+        const algorithmFPType HUGE = 1.0e30;
+
+        algorithmFPType minVal = HUGE;
+        int minIdx             = -1;
+        for (int i = 0; i < numgrp; i++)
+        {
+            algorithmFPType curVal   = (i == numgrp - 1 && local_id >= rem) ? HUGE : distances[global_id + N * (local_id + i * size)];
+            algorithmFPType localMin = sub_group_reduce_min(curVal);
+            if (localMin < minVal)
+            {
+                minVal = localMin;
+                minIdx = i;
+            }
+        }
+        int curInd = 1;
+        if (minIdx < numgrp - 1 || local_id < rem) curInd = (distances[global_id + N * (local_id + minIdx * size)] > minVal) ? 1 : -local_id;
+        int ind = minIdx * size - sub_group_reduce_min(curInd);
+        if (local_id == 0)
+        {
+            assignments[global_id]  = ind;
+            mindistances[global_id] = minVal;
+        }
+    }
+
+    int insert_subgroup_shift_right(__local int * data, int index, int offset, int newVal, int sub_group_id, int sub_group_size, int rem) {
+        if (index < 0) index = 0;
+        int ret     = data[offset + sub_group_size - 1];
+        int curVal  = data[offset + sub_group_id];
+        int prevVal = curVal;
+        int delta   = index < sub_group_id ? 1 : 0;
+        int res     = intel_sub_group_shuffle_up(prevVal, curVal, delta);
+        if (rem == 0 || sub_group_id < rem)
```

Comment:
i recommend using condition style "if(0 == val" for such checks to escape wasting time on finding misprint issues like if(val =0